File: whisper/src/main/java/com/dmrandevu/whisper/WhisperContext.kt
package com.dmrandevu.whisper

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import java.io.Closeable
import java.io.File
import java.util.concurrent.Executors

/**
 * A loaded whisper model.
 *
 * Everything runs on one dedicated thread. whisper's context holds decoding state that cannot be
 * touched from two places at once, and the alternative — a lock around every call — would still
 * leave the abort flag racing with whichever pass happened to be running.
 */
class WhisperContext private constructor(private var contextPtr: Long) : Closeable {

    class TranscriptionFailedException(message: String) : Exception(message)

    /**
     * One segment as whisper reported it. In [noTimestamps] mode there is only ever one.
     *
     * [alignedStartMs] is where cross-attention alignment puts it, which is the one to trust;
     * [startMs] comes from the decoder's own timestamp tokens and collapses onto window
     * boundaries. Null when the model was loaded without alignment.
     */
    data class Segment(
        val text: String,
        val startMs: Long,
        val endMs: Long,
        val alignedStartMs: Long? = null
    ) {
        /** The best timing available for this segment. */
        val bestStartMs: Long get() = alignedStartMs ?: startMs
    }

    /**
     * Which set of attention heads alignment should read. Fixed per model architecture — the
     * wrong one produces confident nonsense rather than an error.
     */
    enum class Alignment(val preset: Int) {
        NONE(0),
        BASE(6),
        SMALL(8)
    }

    /**
     * Transcribes 16 kHz mono samples.
     *
     * See [WhisperLib.fullTranscribe] for what [noTimestamps] actually changes — it is not a
     * printing option.
     */
    suspend fun transcribe(
        samples: FloatArray,
        noTimestamps: Boolean,
        threads: Int = defaultThreads(),
        beamSize: Int = DEFAULT_BEAM_SIZE,
        noContext: Boolean = false
    ): List<Segment> = withContext(worker) {
        currentCoroutineContext().ensureActive()
        check(contextPtr != 0L) { "Model already closed" }

        val result = WhisperLib.fullTranscribe(contextPtr, threads, samples, noTimestamps, beamSize, noContext)
        when (result) {
            0 -> Unit
            ABORTED -> {
                // The caller's coroutine is what set the flag, so let its cancellation surface.
                currentCoroutineContext().ensureActive()
                throw TranscriptionFailedException("Recognition was aborted")
            }

            else -> throw TranscriptionFailedException("Recognition failed with code $result")
        }

        (0 until WhisperLib.segmentCount(contextPtr)).map { index ->
            val aligned = WhisperLib.segmentAlignedStart(contextPtr, index)
            Segment(
                text = WhisperLib.segmentText(contextPtr, index),
                // whisper counts in hundredths of a second.
                startMs = WhisperLib.segmentStart(contextPtr, index) * 10,
                endMs = WhisperLib.segmentEnd(contextPtr, index) * 10,
                alignedStartMs = if (aligned >= 0) aligned * 10 else null
            )
        }
    }

    /**
     * Stops a transcription that is already running.
     *
     * `whisper_full` blocks for tens of seconds, and it holds the worker thread while it does —
     * so cancelling the coroutine alone would leave a core busy long after the operator moved on.
     */
    fun abort() = WhisperLib.setAbort(true)

    override fun close() {
        if (contextPtr == 0L) return
        WhisperLib.freeContext(contextPtr)
        contextPtr = 0
    }

    companion object {
        private const val ABORTED = -2

        /**
         * Matches whisper.cpp's own command line tool. Greedy decoding is quicker but its token
         * timestamps collapse onto the 30-second window boundaries, which puts the beep before
         * the word rather than over it.
         */
        const val DEFAULT_BEAM_SIZE = 5

        /** Serialises every call into the native context; see the class comment. */
        private val worker = Executors.newSingleThreadExecutor { runnable ->
            Thread(runnable, "whisper").apply { isDaemon = true }
        }.asCoroutineDispatcher()

        suspend fun load(
            model: File,
            alignment: Alignment = Alignment.NONE
        ): WhisperContext = withContext(worker) {
            require(model.exists()) { "Model file missing: ${model.absolutePath}" }
            val ptr = WhisperLib.initContext(model.absolutePath, alignment.preset)
            if (ptr == 0L) throw TranscriptionFailedException("Could not load ${model.name}")
            WhisperContext(ptr)
        }

        fun systemInfo(): String = WhisperLib.systemInfo()

        /**
         * Leaves a couple of cores alone. Recognition already makes the operator wait; taking
         * every core with it would stall the video playing behind the progress counter too.
         */
        private fun defaultThreads() =
            (Runtime.getRuntime().availableProcessors() - 2).coerceIn(2, 6)
    }
}

File: ios/DMRandevuGaleri/Media/Censor/WhisperContext.swift
import Foundation

/// A loaded whisper model.
///
/// Everything runs on one dedicated queue. whisper's context holds decoding state that cannot be
/// touched from two places at once, and the alternative — a lock around every call — would still
/// leave the abort flag racing with whichever pass happened to be running.
actor WhisperContext {

    struct TranscriptionFailedError: Error {
        var message: String
    }

    /// One segment as whisper reported it. With `maxLen` 1 there is one token in each.
    struct Segment {
        var text: String
        var startMs: Int64
        var endMs: Int64
    }

    private var pointer: OpaquePointer?

    private init(pointer: OpaquePointer) {
        self.pointer = pointer
    }

    deinit {
        if let pointer { whisper_free(pointer) }
    }

    static func load(model: URL) throws -> WhisperContext {
        guard FileManager.default.fileExists(atPath: model.path) else {
            throw TranscriptionFailedError(message: "Model file missing: \(model.lastPathComponent)")
        }
        var params = whisper_context_default_params()
        // Metal, which is why this is worth doing on the phone at all. Unlike the Android build
        // there is no reason to stay on the CPU here: nothing in this port asks for the
        // cross-attention alignment that would force flash attention off.
        params.use_gpu = true
        guard let pointer = whisper_init_from_file_with_params(model.path, params) else {
            throw TranscriptionFailedError(message: "Could not load \(model.lastPathComponent)")
        }
        return WhisperContext(pointer: pointer)
    }

    /// Transcribes 16 kHz mono samples.
    ///
    /// `noTimestamps` is not a printing option — it changes what the decoder produces. Suppressed,
    /// whisper transcribes swearing it otherwise replaces with an innocent near-homophone, but it
    /// answers with one thirty-second block and cannot say when. With `maxLen` 1 it emits a
    /// segment per token, which is where word timings come from. Both run, and the results are
    /// lined up afterwards.
    func transcribe(
        samples: [Float],
        noTimestamps: Bool,
        maxLen: Int32? = nil,
        beamSize: Int32 = WhisperContext.defaultBeamSize
    ) throws -> [Segment] {
        guard let pointer else {
            throw TranscriptionFailedError(message: "Model already closed")
        }
        try Task.checkCancellation()

        var params = whisper_full_default_params(
            beamSize > 1 ? WHISPER_SAMPLING_BEAM_SEARCH : WHISPER_SAMPLING_GREEDY
        )
        if beamSize > 1 { params.beam_search.beam_size = beamSize }
        params.print_realtime = false
        params.print_progress = false
        params.print_timestamps = false
        params.print_special = false
        params.translate = false
        params.n_threads = Int32(Self.threadCount)
        params.offset_ms = 0
        params.single_segment = false
        params.no_timestamps = noTimestamps
        params.max_len = maxLen ?? (noTimestamps ? 0 : 1)
        params.token_timestamps = !noTimestamps
        // Every call stands alone, and must. whisper otherwise primes a call with the text of the
        // one before it, and this context is reused across passes over the same clip: on Android
        // a short snippet decoded after three full passes came back with different words and its
        // timestamps piled onto one edge of the audio, which looked like the hardware being
        // unreliable and was this.
        params.no_context = true

        let language = Self.language
        let result: Int32 = language.withCString { pointer2 in
            params.language = pointer2
            return samples.withUnsafeBufferPointer { audio in
                whisper_full(pointer, params, audio.baseAddress, Int32(audio.count))
            }
        }
        guard result == 0 else {
            throw TranscriptionFailedError(message: "Recognition failed with code \(result)")
        }

        return (0..<whisper_full_n_segments(pointer)).map { index in
            Segment(
                text: String(cString: whisper_full_get_segment_text(pointer, index)),
                // whisper counts in hundredths of a second.
                startMs: Int64(whisper_full_get_segment_t0(pointer, index)) * 10,
                endMs: Int64(whisper_full_get_segment_t1(pointer, index)) * 10
            )
        }
    }

    func close() {
        if let pointer { whisper_free(pointer) }
        pointer = nil
    }

    private static let language = "tr"

    /// Matches whisper.cpp's own command line tool.
    static let defaultBeamSize: Int32 = 5

    /// Leaves a couple of cores alone. Recognition already makes the operator wait; taking every
    /// core with it would stall the video playing behind the progress counter too.
    private static var threadCount: Int {
        max(2, min(6, ProcessInfo.processInfo.activeProcessorCount - 2))
    }
}
